In [1]:
#Run the following code to print multiple outputs from a cell
get_ipython().ast_node_interactivity = 'all'

# Profiling

The data for today comes from the FAA's Wildlife Strike Database (https://wildlife.faa.gov/home). The purpose of this database is to study the frequency of wildlife strikes causing damage to aircraft, as well as to track both injuries and fatalities caused by such strikes.

## Import Data and Initial Checks
Import the file, "StrikeReportsPartial.csv", and save it in a variable called `df`. Then, do some initial checks of the data based on what we covered last class. What can you say about the data?

In [2]:
import pandas
df = pandas.read_csv("StrikeReportsPartial.csv")
df
df.dtypes
df.describe()

INDEX_NR OPID           ATYPE  AC_MASS INCIDENT_DATE TIME_OF_DAY  \
0         260720  PHM   SIKORSKY S-76      2.0    2009-01-04         Day   
1         260583  BUS        CITATION      2.0    2008-03-04         Day   
2         128423  PVT    PA-23 APACHE      1.0    1998-03-04         Day   
3         221264  BUS           C-172      1.0    2003-07-08         Day   
4         251781  BUS  PA-44 SEMINOLE      2.0    2007-10-23       Night   
...          ...  ...             ...      ...           ...         ...   
157522    358804  UNK         UNKNOWN      NaN    2015-03-30         NaN   
157523    358805  UNK         UNKNOWN      NaN    2015-03-31         NaN   
157524    358806  UNK         UNKNOWN      NaN    2015-03-31         NaN   
157525    358807  UNK         UNKNOWN      NaN    2015-03-31         NaN   
157526    358808  UNK         UNKNOWN      NaN    2015-03-31         NaN   

       AIRPORT_ID  SPEED  DISTANCE PHASE_OF_FLT DAMAGE         SKY PRECIP  \
0            UNKN  135.0       NaN     En Route      D  Some Cloud    NaN   
1            KPWA    NaN       4.0        Climb      D    No Cloud    NaN   
2            UNKN  146.0       NaN     En Route      D    No Cloud    NaN   
3            UNKN  110.0       NaN     En Route      D  Some Cloud    NaN   
4            UNKN  150.0       NaN     En Route      D    No Cloud    NaN   
...           ...    ...       ...          ...    ...         ...    ...   
157522       KUES    NaN       0.0          NaN    NaN         NaN    NaN   
157523       KCMA    NaN       0.0          NaN    NaN         NaN    NaN   
157524       KMCI    NaN       0.0          NaN    NaN         NaN    NaN   
157525       KMFE    NaN       0.0          NaN    NaN         NaN    NaN   
157526       KMIV    NaN       0.0          NaN    NaN         NaN    NaN   

        NR_INJURIES  NR_FATALITIES  NR_INJURIES_0  NR_FATALITIES_0  
0               1.0            8.0              1                8  
1               NaN            5.0              0                5  
2               NaN            2.0              0                2  
3               NaN            2.0              0                2  
4               NaN            2.0              0                2  
...             ...            ...            ...              ...  
157522          NaN            NaN              0                0  
157523          NaN            NaN              0                0  
157524          NaN            NaN              0                0  
157525          NaN            NaN              0                0  
157526          NaN            NaN              0                0  

[157527 rows x 17 columns]

INDEX_NR             int64
OPID                object
ATYPE               object
AC_MASS            float64
INCIDENT_DATE       object
TIME_OF_DAY         object
AIRPORT_ID          object
SPEED              float64
DISTANCE           float64
PHASE_OF_FLT        object
DAMAGE              object
SKY                 object
PRECIP              object
NR_INJURIES        float64
NR_FATALITIES      float64
NR_INJURIES_0        int64
NR_FATALITIES_0      int64
dtype: object

INDEX_NR        AC_MASS         SPEED      DISTANCE  NR_INJURIES  \
count  157527.000000  119586.000000  63387.000000  90694.000000   226.000000   
mean   246417.848185       3.517360    141.638680      0.658376     1.318584   
std     78440.809372       0.873351     46.019708      3.307637     0.813940   
min    100000.000000       1.000000      0.000000      0.000000     1.000000   
25%    209260.500000       3.000000    120.000000      0.000000     1.000000   
50%    248788.000000       4.000000    136.000000      0.000000     1.000000   
75%    318105.500000       4.000000    160.000000      0.000000     1.000000   
max    358808.000000       5.000000    380.000000    150.000000     7.000000   

       NR_FATALITIES  NR_INJURIES_0  NR_FATALITIES_0  
count      12.000000  157527.000000    157527.000000  
mean        2.166667       0.001892         0.000165  
std         2.167249       0.058627         0.026183  
min         1.000000       0.000000         0.000000  
25%         1.000000       0.000000         0.000000  
50%         1.000000       0.000000         0.000000  
75%         2.000000       0.000000         0.000000  
max         8.000000       7.000000         8.000000

## Missing Data

Summary statistics don't help us learn much about missing data. Missing data is often an important consideration because it is not missing at random. 

The `.isna()` method tells us if a value is missing. Run the following command and compare it to the initial preview of the data above:

In [3]:
df.isna()

INDEX_NR   OPID  ATYPE  AC_MASS  INCIDENT_DATE  TIME_OF_DAY  \
0          False  False  False    False          False        False   
1          False  False  False    False          False        False   
2          False  False  False    False          False        False   
3          False  False  False    False          False        False   
4          False  False  False    False          False        False   
...          ...    ...    ...      ...            ...          ...   
157522     False  False  False     True          False         True   
157523     False  False  False     True          False         True   
157524     False  False  False     True          False         True   
157525     False  False  False     True          False         True   
157526     False  False  False     True          False         True   

        AIRPORT_ID  SPEED  DISTANCE  PHASE_OF_FLT  DAMAGE    SKY  PRECIP  \
0            False  False      True         False   False  False    True   
1            False   True     False         False   False  False    True   
2            False  False      True         False   False  False    True   
3            False  False      True         False   False  False    True   
4            False  False      True         False   False  False    True   
...            ...    ...       ...           ...     ...    ...     ...   
157522       False   True     False          True    True   True    True   
157523       False   True     False          True    True   True    True   
157524       False   True     False          True    True   True    True   
157525       False   True     False          True    True   True    True   
157526       False   True     False          True    True   True    True   

        NR_INJURIES  NR_FATALITIES  NR_INJURIES_0  NR_FATALITIES_0  
0             False          False          False            False  
1              True          False          False            False  
2              True          False          False            False  
3              True          False          False            False  
4              True          False          False            False  
...             ...            ...            ...              ...  
157522         True           True          False            False  
157523         True           True          False            False  
157524         True           True          False            False  
157525         True           True          False            False  
157526         True           True          False            False  

[157527 rows x 17 columns]

There are 157,527 rows of data in this file, so it's not very helpful to see the `True` and `False` for every column and every row. Instead, let's summarize with the `.sum()` method:

In [4]:
df.isna().sum()

INDEX_NR                0
OPID                    0
ATYPE                   0
AC_MASS             37941
INCIDENT_DATE           0
TIME_OF_DAY         58690
AIRPORT_ID              0
SPEED               94140
DISTANCE            66833
PHASE_OF_FLT        51766
DAMAGE              44859
SKY                 71916
PRECIP             150027
NR_INJURIES        157301
NR_FATALITIES      157515
NR_INJURIES_0           0
NR_FATALITIES_0         0
dtype: int64

*In python, the value of `True` is treated as a `1` for the purposes of the `.sum()` method.*

Do we have a missing values problem? What do you think is going on here?

### `pandas` formatting side note

Take a moment to look at the last line of code you ran (`df.isna().sum()`). We're going to be using this type of formatting alot where you list the name of a DataFrame and then list commands separated by `.`'s. Can you guess what is going on here?

Here's another example of this...before running the command, what do you think this will do?

In [5]:
df["TIME_OF_DAY"].isna().sum()

58690

## Accessing Individual Columns

Up to now, we've worked with all columns. But often we want to focus on a specific column. Python uses brackets to restrict a DataFrame to one (or more) columns. For example:

In [6]:
df["TIME_OF_DAY"]

0           Day
1           Day
2           Day
3           Day
4         Night
          ...  
157522      NaN
157523      NaN
157524      NaN
157525      NaN
157526      NaN
Name: TIME_OF_DAY, Length: 157527, dtype: object

Now you try...access the column, DAMAGE:

In [7]:
df["DAMAGE"]

0           D
1           D
2           D
3           D
4           D
         ... 
157522    NaN
157523    NaN
157524    NaN
157525    NaN
157526    NaN
Name: DAMAGE, Length: 157527, dtype: object

What if you want more than a single column? Just put a list inside the brackets. In python, a list is also enclosed in brackets, with the elements separated by commas. So, if you wanted both TIME_OF_DAY and DAMAGE, you'd run this command:

In [8]:
df[["TIME_OF_DAY", "DAMAGE"]]

TIME_OF_DAY DAMAGE
0              Day      D
1              Day      D
2              Day      D
3              Day      D
4            Night      D
...            ...    ...
157522         NaN    NaN
157523         NaN    NaN
157524         NaN    NaN
157525         NaN    NaN
157526         NaN    NaN

[157527 rows x 2 columns]

Display the ATYPE and AC_MASS columns:

In [9]:
df[["ATYPE", "AC_MASS"]]

ATYPE  AC_MASS
0        SIKORSKY S-76      2.0
1             CITATION      2.0
2         PA-23 APACHE      1.0
3                C-172      1.0
4       PA-44 SEMINOLE      2.0
...                ...      ...
157522         UNKNOWN      NaN
157523         UNKNOWN      NaN
157524         UNKNOWN      NaN
157525         UNKNOWN      NaN
157526         UNKNOWN      NaN

[157527 rows x 2 columns]

Now, with those same 2 columns, sum up the number of missing values:

In [10]:
df[["ATYPE", "AC_MASS"]].isna().sum()

ATYPE          0
AC_MASS    37941
dtype: int64

The NR_INJURIES column contains numeric data. Answer the following 3 questions:
* How many missing values are there in the NR_INJURIES column?
* Add up the total number of injuries in this DataFrame
* Get the average number (or `mean()`) of injuries in this DataFrame

In [11]:
df["NR_INJURIES"].isna().sum()
df["NR_INJURIES"].sum()
df["NR_INJURIES"].mean().round(3)

157301

298.0

1.319

The 3 most common descriptive statistics methods for numeric data in pandas are: `.sum()`, `.mean()`, and `.std()`.

## Summarizing Categorical Data: Frequency Tables

The `.value_counts()` method lets you know how frequently values appear in a column.

In [12]:
df["TIME_OF_DAY"].value_counts()

TIME_OF_DAY
Day      60963
Night    30164
Dusk      4355
Dawn      3353
NIGHT        1
DAY          1
Name: count, dtype: int64

By the way, what data quality issue can we see based on this frequency table?

Now, build a frequency table for the DAMAGE column:

In [13]:
df["DAMAGE"].value_counts()

DAMAGE
N     98523
M      7553
S      3775
M?     2751
D        66
Name: count, dtype: int64

If you want the percentage/proportion instead of the count, add the `normalize=True` parameter to `.value_counts()`. Try it with the DAMAGE column:

In [14]:
df["DAMAGE"].value_counts(normalize = True)

DAMAGE
N     0.874454
M     0.067038
S     0.033506
M?    0.024417
D     0.000586
Name: proportion, dtype: float64

By default, python ignores missing values (`.value_counts()` has a parameter `dropna` which defaults to `True`). If you'd like the percentage relative to all the rows, including those with missing values, set `dropna` equal to `False`:

In [15]:
df["DAMAGE"].value_counts(normalize = True, dropna = False)

DAMAGE
N      0.625436
NaN    0.284770
M      0.047947
S      0.023964
M?     0.017464
D      0.000419
Name: proportion, dtype: float64

Now, use the `round()` function on the previous command to limit the decimals to 2 places:

In [16]:
df["DAMAGE"].value_counts(normalize = True, dropna = False).round(2)

DAMAGE
N      0.63
NaN    0.28
M      0.05
S      0.02
M?     0.02
D      0.00
Name: proportion, dtype: float64

Now, you try...
* What % of incidents happened at Dusk (from TIME_OF_DAY)...*ignoring* missing rows?
* What % of incidents happened at Dusk (from TIME_OF_DAY)...*including* missing rows?

Use the `round()` function to round all numbers to 3 decimals.

## Contingency Tables

Frequency tables show us counts/percentages for a single column. Contingency tables show the counts/percentages when combining 2 columns. For this, we can use the `crosstab()` function from the `pandas` module:

In [ ]:
pandas.crosstab(df["DAMAGE"], df["TIME_OF_DAY"])

Notice this is a *function* and not a *method*, so the name of the function comes first and then the first 2 parameters are the names of the columns you want to include.

If you want to also display row and column totals, include the `margins=True` parameter:

In [ ]:
pandas.crosstab(df["DAMAGE"], df["TIME_OF_DAY"], margins = True)

Create a contingency table for TIME_OF_DAY and PRECIP, including row and column totals:

You can also get percentages/proportions using the `normalize` parameter. For the `crosstab()` function, the `normalize` parameter has 3 possible values: `"all"`, `"index"`, and `"columns"`. 

The `"all"` setting gives percentages out of the total:

In [ ]:
pandas.crosstab(df["DAMAGE"], df["TIME_OF_DAY"], margins = True, normalize = "all")

However, usually it's more helpful to %'s by row or column.

The `"index"` setting gives row %'s and the `"columns"` setting gives column %'s:

In [ ]:
pandas.crosstab(df["DAMAGE"], df["TIME_OF_DAY"], margins = True, normalize = "index")
pandas.crosstab(df["DAMAGE"], df["TIME_OF_DAY"], margins = True, normalize = "columns")

# Basic Visualization

`pandas` supports visualizations, but requires an additional module, `matplotlib`. This module has already been installed on Jupyter Hub and will automatically load whenever you `import pandas`.

## Summarizing Numeric Data

To summarize numeric data visually, ***histograms*** and ***box plots*** are usually your first best choices.

### Histograms

Histograms give a good sense of how your data is distributed. Run the following command...what can you tell about the number of injuries in this DataFrame?

In [ ]:
df["NR_INJURIES"].plot.hist()

Now create the same histogram, but this time using the NR_INJURIES_0 column. (This is a column I included to replace missing values with a 0.)

Comparing both histograms, what story does this tell?

In [ ]:
df["NR_INJURIES_0"].plot.hist()

### Box Plots

The box plot command also has a similar structure to histograms:

In [ ]:
df["SPEED"].plot.box()

The green line represents the median (or 50%) point, with the bottom and top of the box representing the 25% and 75% points respectively. In other words, 50% of the data is within the range shown by the box. The points shown above and below the "whiskers" are considered outliers.

## Summarizing Categorical Data

To summarize categorical data visually, you should use a ***bar chart***. Although bar charts look similar to histograms, they summarize data differently. A histogram takes continuous numerical data and counts the frequency of observations within a certain range (or bin). Bar charts, on the other hand, count up the number of observations in each category. You should use bar charts *only* for categorical data and histograms for numeric data.

### Bar Charts

Bar charts are good for visually summarizing categorical data. Run the following command to see what time of day most incidents occur on:

In [ ]:
df["TIME_OF_DAY"].plot.bar()

What happened? The `.plot.bar()` method needs summary data to visualize. Can you figure out how to add to the code above to get the bar chart to display?

In [ ]:
df["TIME_OF_DAY"].value_counts().plot.bar()

Just like with frequency tables, so too with bar charts...you can also set the y-axis to show percentages instead of counts:

In [ ]:
df["TIME_OF_DAY"].value_counts(normalize = True).plot.bar()

## Visualizing Relationships

To visually depict the relationship between 2 numeric variables, you can use a ***scatter plot***. Scatter plots should only be used with numeric data. If you'd like to compare data across multiple categories, you can use bar charts or side-by-side box plots.

### 2 Numeric Variables: Scatter Plots

Because scatterplots show the relationship between 2 numeric variables, the structure of the command is slightly different than the basic plots above:

In [ ]:
df.plot.scatter(x = "AC_MASS", y = "NR_INJURIES")

### 1 Numeric & 1 Categorical Variable: Side-by-Side Box Plots

If you want to see how the distribution of a numeric variable varies across categories, you can use side-by-side box plots.

The command for a side-by-side box plot is different than for the basic box plot, but similar in structure to scatter plots:

In [ ]:
df.boxplot(column = "SPEED", by = "DAMAGE")

The `column` parameter contains the numeric variable you are plotting and the `by` parameter contains a categorical variable you'd like to split the box plots by.

### 2 Categorical Variables: Stacked Bar Charts

To visualize the relationship between 2 categorical variables, you can use a stacked bar chart.

Remember that a bar chart for 1 variable is just a visualization of a frequency table (using `.value_counts()`). A bar chart for 2 variables is a visualization of a contingency table, created with the `crosstab()` function:

In [ ]:
pandas.crosstab(df["DAMAGE"], df["TIME_OF_DAY"]).plot.bar(stacked = True)

As you can see, when there is an imbalance in data across categories, simply looking at the counts can make comparison difficult. One way to fix this is to add `normalize = "index"` to your `crosstab()` command to depict percentages instead:

In [ ]:
pandas.crosstab(df["DAMAGE"], df["TIME_OF_DAY"], normalize = "index").plot.bar(stacked = True)

## Visualization Practice

Create the following charts:

In [ ]:
# Make a histogram looking at the number of fatalities (NR_FATALITIES)


In [ ]:
# Make a percentage bar chart by DAMAGE type


In [ ]:
# Make a scatterplot showing the relationship between SPEED and DISTANCE


In [ ]:
# Make a box plot looking at DISTANCE broken out by TIME_OF_DAY


## Data Profiling Summary

Typically, when first reading in your data, it's a good idea to do the following to profile the data:
* Look at the shape of your DataFrame
* Preview some rows
* Get a sense of the data types for each column
* Use `.isna().sum()` to see how many missing values there are
* Use `.describe()` to summarize the numeric variables
* Use `.value_counts()` and/or `crosstab()` to summarize the categorical variables
* Create charts to quickly visualize your data
   * Histograms and Box Plots are good for visualizing numeric data
   * Bar Charts are good for visualizing categorical data
   * Scatterplots are good for comparing 2 numeric variables to identify relationships